### PR TITLE
First pass at metadata support for new trees

### LINKED ISSUE #2651: Scan filters trie pages using content metadata

### PR DESCRIPTION
**Pull request by FiV0**
This takes a first pass at metadata support for the new trees. No pushdown blooms for now.

Open question:
- [x] Should the metadata expression namespace be updated for boolean metadata? Moved to it's own ticket #2717 
- [x] ~Good naming for `chunk-idx/trie-idx`~

### CODE REVIEW

File: core/src/main/clojure/xtdb/metadata.clj [hunk 1]
```diff
@@ -270,8 +268,56 @@
...
+                                    (not (nil? (.getObject page-idx-rdr idx))) inc)
+                                  (inc idx))))}))
+
+(defn- table-name->dir [table-name] (format "tables/%s/chunks" table-name))
+
+(defn ->table-leaf-obj-key [table-name chunk-idx]
+  (format "%s/leaf-c%s.arrow" (table-name->dir table-name) chunk-idx))
+
+(defn ->table-trie-obj-key [table-name chunk-idx]
+  (format "%s/trie-c%s.arrow" (table-name->dir table-name) chunk-idx))
```
Comment (FiV0) [inline core/src/main/clojure/xtdb/metadata.clj:285]: Would have preferred for the above 3 to stay in `xtdb.trie` but that would have created a cyclic dep.

File: core/src/main/clojure/xtdb/metadata.clj [hunk 2]
```diff
@@ -281,16 +325,23 @@
...
-    #_
-    (with-single-root buffer-pool (->table-metadata-obj-key chunk-idx table-name)
-      (fn [metadata-root]
-        (.apply f (->table-metadata metadata-root
-                                    (.computeIfAbsent table-metadata-idxs
-                                                      [chunk-idx table-name]
-                                                      (reify Function
-                                                        (apply [_ _]
-                                                          (->table-metadata-idxs metadata-root)))))))))
+    (CompletableFuture/completedFuture
```
Comment (FiV0) [inline core/src/main/clojure/xtdb/metadata.clj]: Should this be somehow become an async call?

File: core/src/main/clojure/xtdb/expression/metadata.clj [hunk 3]
```diff
@@ -106,26 +107,31 @@
...
+(def ^:private bloom-rdr-sym (gensym "bloom-rdr"))
+
 
 (defmethod expr/codegen-expr :metadata-vp-call [{:keys [f meta-value field param-expr col-type bloom-hash-sym]} opts]
   (let [field-name (util/str->normal-form-str (str field))
 
-        idx-code `(.rowIndex ~table-metadata-sym ~field-name ~block-idx-sym)]
+        idx-code `(.rowIndex ~table-metadata-sym ~field-name ~page-idx-sym)]
+
+    (prn meta-value)
```
Comment (jarohen) [inline core/src/main/clojure/xtdb/expression/metadata.clj]: cheeky prn :wink:

File: core/src/main/clojure/xtdb/expression/metadata.clj [hunk 4]
```diff
@@ -23,16 +20,20 @@
...
   (letfn [(var-param-expr [f meta-value field {:keys [param-type] :as param-expr}]
-            (let [base-col-types (-> (get col-types field)
-                                     types/flatten-union-types)]
-              (simplify-and-or-expr
-                {:op :call
-                 :f :or
-                 ;; TODO this seems like it could make better use
-                 ;; of the polymorphic expr patterns?
-                 :args (vec
+            ;; TODO adapt for boolean metadata writer
```
Comment (jarohen) [inline core/src/main/clojure/xtdb/expression/metadata.clj:27]: cheers for raising the other card - not sure I entirely understand its impact, shall we discuss tomorrow?

File: core/src/main/clojure/xtdb/expression/metadata.clj [hunk 5]
```diff
@@ -106,26 +107,31 @@
...
 (defmethod expr/codegen-expr :metadata-vp-call [{:keys [f meta-value field param-expr col-type bloom-hash-sym]} opts]
   (let [field-name (util/str->normal-form-str (str field))
 
-        idx-code `(.rowIndex ~table-metadata-sym ~field-name ~block-idx-sym)]
+        idx-code `(.rowIndex ~table-metadata-sym ~field-name ~page-idx-sym)]
+
+    (prn meta-value)
 
-    (if (= meta-value :bloom-filter)
+    (cond
```
Comment (jarohen) [inline core/src/main/clojure/xtdb/expression/metadata.clj]: `if` -> `cond`? looks like it still only has one case?

File: core/src/main/clojure/xtdb/expression/metadata.clj [hunk 6]
```diff
@@ -141,9 +147,8 @@
                                    (assoc-in [:var->col-type col-sym] (types/merge-col-types col-type :null))))]
         {:return-type :bool
          :batch-bindings [[(-> col-sym (expr/with-tag IVectorReader))
-                           `(some-> ^StructVector (.getChild ~types-vec-sym ~(.getName col-field))
-                                    (.getChild ~(name meta-value))
-                                    vr/vec->reader)]]
+                           `(some-> ^IVectorReader (.structKeyReader ~types-rdr-sym ~(.getName col-field))
```
Comment (jarohen) [inline core/src/main/clojure/xtdb/expression/metadata.clj]: nit: shouldn't need type-hints here now?

File: core/src/main/clojure/xtdb/metadata.clj [hunk 7]
```diff
@@ -270,8 +266,56 @@
             (write-col-meta! true col))
           (.endList cols-wtr))))))
 
+(defn ->table-metadata-idxs [^IVectorReader metadata-rdr]
+  (let [page-idx-cache (HashMap.)
+        meta-row-count (.valueCount metadata-rdr)
+        ^IVectorReader #_<Int> page-idx-rdr (.structKeyReader metadata-rdr "page-idx")
```
Comment (jarohen) [inline core/src/main/clojure/xtdb/metadata.clj]: nit: shouldn't need type hints all the way down here either

File: core/src/main/clojure/xtdb/metadata.clj (hunk 2 above)
Comment (jarohen) [reply core/src/main/clojure/xtdb/metadata.clj]: eventually, sure. might be as straightforward as first doing the `(.getBuffer ...)` call, then calling `util/then-apply` with the rest of the body. you'll also want to ensure the buffer's closed - think a try/finally within the then-apply should suffice this time?

File: core/src/main/clojure/xtdb/metadata.clj [hunk 8]
```diff
@@ -333,47 +384,41 @@
...
-                (when (.test pred block-idx)
-                  (.add block-idxs block-idx)))
-
-              (when-not (.isEmpty block-idxs)
-                (->ChunkMatch chunk-idx block-idxs (.columnNames table-metadata))))))))))
-
+     (fn [^long trie-idx, ^ITableMetadata table-metadata]
+       (let [pred (.build metadata-pred table-metadata)
+             page-idxs (RoaringBitmap.)]
+         (prn [trie-idx (.pageCount table-metadata)])
```
Comment (jarohen) [inline core/src/main/clojure/xtdb/metadata.clj]: cheeky prn :wink:

**Review by jarohen — CHANGES_REQUESTED** [review]
cheers @FiV0, well done :clap: 

Best to have a walkthrough of this one, I suspect - there's quite a bit going on :sweat_smile: Particularly, I'd like to understand the separate `trie-idx`, `chunk-idx` has always been sufficient for this so far - but of course requirements have changed quite a bit since then!

File: core/src/main/clojure/xtdb/metadata.clj (hunk 2 above)
Comment (FiV0) [reply core/src/main/clojure/xtdb/metadata.clj]: An error from the `getBuffer` call would then properly get propagated? Looking at `util/rethrowing-cause` wrapping.

**Pull request merged**